Add job registry; remove task registry; remove `Job#to_enum`

benarmston opened the pull request:
This PR adds a `JobRegistry`, removes `TaskRegistry` and removes the `Job#to_enum` abstraction.

The job registry maintains a registry of all jobs regardless of job type.  It provides a means to retrieve a job by ID and also to retrieve all registered tasks for an array job.

Previously, a list of running tasks for an array job used to be maintained by the `TaskRegistry` this is now queried from the `JobRegistry`.  This is done by querying all tasks for the job and filtering for any that are `RUNNING`.

Previously, `TaskRegistry` was used to obtain the next pending task for an array job.  Now there is an `ArrayTaskGenerator` class which deals soley with generating the next pending task and querying the current state of the task generation.  In particular, it doesn't maintain a list of tasks grouped by state, or know when the next pending task is no longer pending.  When a pending task is scheduled, the scheduler informs `ArrayTaskGenerator` of this fact.

`FifoScheduler` no longer maintains its own queue of jobs.  Instead it filters and sorts the jobs present in the job registry.  This is sufficient for the `FifoScheduler`, but a more complicated scheduler may need to do something more complicated.

---

Previously we had a couple of methods to lookup a job.  Either by filtering through the scheduler's queue or finding the job from its allocation.  All of these methods have been updated to query the job registry for the job.

---

Moving a job into a terminal state has seen some changes.

Previously, the cancellation classes moved the job into a `CANCELLED` state once they had sent the request to cancel they job.  Now, they leave it in `CANCELLING` and rely on `EventProcessor` to update the job when it processes the notification from the daemon.

When the last `ARRAY_TASK` is updated to `COMPLETED`, `CANCELLED` or `FAILED`, its `ARRAY_JOB` is also updated accordingly.

`EventProcessor` is no longer responsible for cleaning up jobs or removing them from the job registry.  Instead it is now responsible only for ensuring that the tasks transition to the correct state at the correct time.

A timer task has been written which runs, by default, every 60 seconds and instructs the job registry to remove any old jobs.  Old jobs are `JOB`s and `ARRAY_JOB`s which are `COMPLETED`, `CANCELLED` or `FAILED`.  `ARRAY_TASK`s are not removed until their `ARRAY_JOB` is removed.

benarmston committed:
Use basic auth for specs

benarmston committed:
Add JobRegistry

It maintains a registry of all jobs regardless of job type.  It also
provides a means to retrieve job tasks for an array job.

Currently, it is required to manually remove the jobs when they are no
longer wanted.  This is done well for `JOB` and `JOB_TASK`s but not for
`ARRAY_JOB`s.  This does not currently affect the scheduling, but it
will need to be addressed prior to job persistence being implemented.

A list of running tasks for an array job used to be maintained by the
`TaskRegistry` this is now queried from the `JobRegistry`.

The `FifoScheduler` now no longer maintains its own queue of jobs.
Instead it filters the jobs present in the job registry.

benarmston committed:
Specs for behaviour of scheduler and array tasks

The previous specs tested the implementation not the behaviour.  Those
specs are not going to very useful for checking that a refactoring has
not broken the scheduler.

benarmston committed:
Refactor fifo scheduler to remove task enum thingy

Remove now (and previously) unused methods from task registry and job.

benarmston committed:
Preserve output of queue command

benarmston committed:
Limit number of iterations to attempt to schedule job

benarmston committed:
Replace `TaskRegistry` with `TaskGenerator`

Much of the task registry functionality has been removed with the
removal of `Job#to_enum` or merged into the new `JobRegistry`.

The remaining functionality is mostly centered around generating the
next task to be scheduled.  This functionality has been moved to a new
`ArrayTaskGenerator` class.

`ArrayTaskGenerator` generates the array tasks and provides methods to
query the state of the current generation.  Unlike `TaskRegistry` it
does not automatically advance its state instead advancing the state is
left to the scheduler as it is the scheduler that best knows if an array
task has been allocated resources.

With that refactoring `TaskRegistry` no longer serves a purposes and has
been removed.

benarmston committed:
Fix job registry deletion issue

`EventProcessor` no longer deletes jobs from the registry or cleans up
after jobs.  Instead it is responsible for updating the state of a job
and any dependent jobs.

`JobRegistry` has gained a new method to garbage collect any
completed/cancelled/failed JOB jobs and ARRAY_JOB jobs.  When it detects
a job that can be garbage collected, it cleans up the job before doing
so.

A new timer task has been written to periodically instruct `JobRegistry`
to garbage collect itself.  This is only ran when the production
webserver is started; it is not ran during tests.

In order to get the timer task working with the same `JobRegistry`
instance that the sinatra app uses, I've monkey patched Falcon to use a
threaded container.  Falcon really has reached the end of its
usefulness, but it can't be removed until we replace `Async::Websocket`
with some other bi-directional socket library.

benarmston committed:
Fix display issue with ARRAY_JOBs

benarmston committed:
Better docs and method names for ArrayTaskGenerator

benarmston committed:
Fix cancellation issues

A job can complete whilst in the `CANCELLING` state.  If this happens,
it should transition to `COMPLETED`.

A `CANCELLING` `ARRAY_JOB` transitions to `CANCELLED` (or possibly
`COMPLETED`, see above paragraph) when the last of its running tasks
reaches a terminal state.

benarmston committed:
Add configuration for period tasks

benarmston committed:
Sane way to conditionally run timer tasks

WilliamMcCumstie commented on app/models/job.rb:
```diff
@@ -218,4 +180,40 @@ def validate_batch_script
+
+    tasks = FlightScheduler.app.job_registry.tasks_for(self)
+    all_finished = tasks
+      .all? { |t| %w(CANCELLED COMPLETED FAILED).include?(t.state) }
```
:panda_face: Generally the following would be cleaner:

```
all_finished = tasks.all? do |task|
  %w(CANCELLED COMPLETED FAILED).include?(task.state)
end
```

benarmston replied:
Fair enough.

WilliamMcCumstie commented on lib/flight_scheduler/application.rb:
```diff
@@ -64,5 +68,22 @@ def config
       @root ||= Pathname.new(__dir__).join('../../').expand_path
     end
+
+    def init_timer_tasks
```
I have a couple of queries about the "timer_task".

Firstly there is a naming issue that "task" is already a psuedo-reserved term for `job-tasks`. I can see the term "timer_task" came from the fact it is a `Concurrent::TimerTask`, however I was a bit confused about this at first. A name reflecting its purpose rather than mode of operation would be clearer. So something along the lines of `job_registry_updater` or very least `job_registry_updater_task`

Secondly I'm not sure about using `Concurrent::TimerTask`. It is a perfectly valid solution if we where using `puma` and everything was threaded, however it seems out of place with `Falcon`/`Async`. A more consistent implementation would be:

```
Async do |task|
  loop do
    task.sleep ...
    # Async supports timeouts as follows, but I'm not sure it is required in this case
    # task.with_timeout(...)  { ... }
    job_registry.remove_old_jobs
  end
end
```

There are a few subtle differences between the implementations, mainly:
* `Concurrent::TimerTask` spawns a monitor/timing thread, where `Async` spawns a fiber
* `Concurrent::TimerTask` spawns a new worker thread each loop, `Async` re-uses the main fiber
* `Concurrent::TimerTask` runs on the `interval` where `Async` introduces an interval shift (`interval + working_time`)

Finally I'm not sure why this particular operation needs a `timeout` or needs the guarantee it runs on a regular interval:
* Using a timeout risks leaving the the `JobRegistry` in an inconsistent state when the `thread` or `fiber` is unceremoniously killed
* `remove_old_jobs` looks like a fast operation, making the `timeout` redundant and the interval shift negligible
* The removal of jobs is not time sensitive so an interval shift is unlikely to cause a problem. The "close enough is good enough" approach

benarmston replied:
You're quite right about the name not being terribly good.  I imagine that this periodic thingy will expand to include other thingies.  Perhaps allocating jobs and perhaps persisting the job registry.  So its purpose probably isn't updating the registry, but running periodic thingies. 

A very important distinction between `TimerTask` and the block you wrote is that `TimerTask` is robust in the face of exceptions.  In particular, an exception thrown by its block will not prevent the task running again when due.  That is the primary reason that I chose it.

A benefit of the consistent interval/timeout is that it gives a guarantee that the task will not be running for at least half the time.  For processor heavy background threads this is a really nice guarantee.  I'm not sure that this is a processor heavy background thread however. I'm also not sure how fast `remove_old_jobs` is going to be.  It will involve deleting jobs from the file system and perhaps eventually writing out accounting information.

On that note, I've just realised that both `BatchScript#write` and `BatchScript#cleanup` are blocking operations.  These do not play nicely with our `Async` usage.  They will need to be fixed, especially if this background thread is to use the main `Async` fiber.

You make a good point about `JobRegistry` and the thread being killed due to a timeout.  The read/write lock will cope just fine.  However, as currently written we risk deleting an array job and leaving its tasks behind.  We should be robust here regardless of the concurrency mechanism.

benarmston replied:
I've made a few changes to `JobRegistry` and I've convinced myself that it is robust even if the periodic processor is timed out whilst it is running.  `JobRegistry` is thread safe so I can't see any issues with it being accessed by a background thread.

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/fifo_scheduler.rb:
```diff
@@ -124,48 +44,89 @@ def allocate_jobs
 
     new_allocations = []
+    iterations = 0
+    max_iterations = 100
```
This looks like an arbitrary number of iterations and I'm not entirely sure why it is required. Notionally the `FIFOScheduler` should keep scheduling jobs until their are no more jobs to be scheduled. This already looks to be the case with:

```
break if candidate.nil?
```

benarmston replied:
Yep, you're right.  This isn't required for a FIFO scheduler.  Though something similar would be required for a backfilling scheduler.

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/fifo_scheduler.rb:
```diff
@@ -124,48 +44,89 @@ def allocate_jobs
-        if next_job.allocation
-          enum.next
-          next
+        if candidate.job_type == 'ARRAY_JOB'
```
Hmmm this section looks a bit odd, AFAICS it could create a busy loop (hence the `max_interval` counter). How about it is merged into the detect section up above? Something like this:

```
candidate = queue.detect do |job|
  next false unless job.pending? && job.allocation.nil?
  job.job_type == 'ARRAY_JOB' ? !job.task_generator.finished? : true
end
```

benarmston replied:
What does `next false` do?  Is that a typo or some Ruby that I've never seen before?

benarmston replied:
The implementation of `queue` already has that check in it, but we are kind of duplicating it anyway....*thinks*.

WilliamMcCumstie replied:
The `next` keyword allows you to specify a value for the current iteration of the block. It's just a habit I've development when the return value of the block is important.

AKA `next` behaves like `return` but in a block context

benarmston replied:
So `next false unless ...` would cause the remainder of the block to be skipped and evaluate to `false` and for the next iteration to run.

Whereas `next unless ...` would cause the remainder of the block to be skipped and evaluate to `nil` and for the next iteration to run.

I suppose that could be useful in some cases.  Thanks for the answer.

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/fifo_scheduler.rb:
```diff
@@ -124,48 +44,89 @@ def allocate_jobs
         else
+          if candidate.job_type == 'ARRAY_TASK'
+            FlightScheduler.app.job_registry.add(candidate)
+            candidate.array_job.task_generator.advance_next_task
```
I'm struggling a bit with the backwards and forwards between `candidate` being a `ARRAY_JOB` or `ARRAY_TASK`. It looks like it is reassigned a few times which requires constant conditional logic to determine what it is. Whilst we in the refactoring mood, this should probably be addressed.

Firstly the `allocate_job` method should be responsible for updating the `AllocationsRegistry`. This is fairly generic code and will probably be required in other schedulers when we get around to implementing them:

```
# NOTE: Regardless of scheduling model, the following section should be the same
# If the allocation fails because no nodes are available, that is by definition a Resource limit
# Some scheduler *may* want to change the reason to Priority, but this can be done with a flag
#
# Otherwise if the allocation is successful, the `array_task` must be added to the JobRegistry and
# advance the `task_generator`. The allocation must also be added to the Allocation registry. These
# requirements will hold true regardless of scheduler, and will likely be duplicated if not abstracted
def allocate_job(job, reason: 'Resources')
    nodes = job.partition.available_nodes_for(job)
    if nodes.nil?
      if job.job_type == 'ARRAY_TASK'
        job.array_job.reason_pending = reason
      else
        job.reason_pending = reason
      end
      return nil
    else
      Allocation.new(job: job, nodes: nodes).tap do |a|
        FlightScheduler.app.allocations.add(a)
        if job.job_type == 'ARRAY_JOB'
          FlightScheduler.app.job_registry.add(job)
          job.task_generator.advance_next_task
        end
      end
    end
  end
end
```

This should allow the entire section to become:

```
candidate = .. # Guaranteed resolved to either a JOB or JOB_TASK above
if candidate.job_type == 'JOB_TASK'
  allocation = allocate_job(candidate.task_generator.next_task)
else
  allocation = allocate_job(job)
end
allocation ? new_allocations << allocation : break
```

benarmston replied:
These look like good ideas.

benarmston replied:
I've incorporated something largely along those lines.

benarmston commented:
Cancelling of pending array jobs is broken on this branch.  It is fixed in https://github.com/openflighthpc/flight-scheduler-controller/pull/33/commits/823b36a0b933089b2b073ab72813b24e8fb6c235 which is part of #33 .

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/fifo_scheduler.rb:
```diff
@@ -124,48 +44,89 @@ def allocate_jobs
+    # are considered prior to later added jobs.
+    FlightScheduler.app.job_registry.jobs
+      .reject { |j| j.job_type == 'ARRAY_JOB' && j.task_generator.finished? }
+      .reject { |j| %w(COMPLETED CANCELLED FAILED).include?(j.state) }
```
The previous two lines could probably be replaced by a `finished?` flag on the `Job`

benarmston replied:
Perhaps.  Though that raises the question "what does it mean for a job to be finshed?".  For `JOB` and `ARRAY_TASK` that is very clear; its in one of its terminal states.

It's not so clear for an `ARRAY_JOB`.  Does it mean that the task generator has finished?  Or does it mean that the job is in one of its terminal states?  Or does it mean both of them must be true?  Or does it mean that either of them must be true?  All four possibilities are possible.

I suspect that whatever one we would select would prove to be the wrong option and a confusing option as often as it proved to be helpful.

WilliamMcCumstie reviewed (commented):
Just have a few comments about various components

benarmston committed:
Incorporate feedback from FifoScheduler review

* `FifoScheduler#allocate_job` is now likely generic across all
  schedulers.

* `FifoScheduler#allocate_jobs` does much less branching on job type.

benarmston committed:
Job registry is more robust

Exceptions raised during `delete` cannot leave the scheduler in an
inconsistent, unrecoverable state.

benarmston committed:
Stylistic change

benarmston committed:
Add a `Job#terminal_state` method

benarmston committed:
Batch script reading/writing doesn't block async

benarmston commented:
@WilliamMcCumstie can you cast your eyes over this again please?

WilliamMcCumstie commented on config/boot.rb:
```diff
@@ -37,3 +37,7 @@
 require_relative '../app/models/job_step'
+
+unless FlightScheduler.env.test?
+  FlightScheduler.app.init_periodic_processor
```
:panda_face: This can probably be launched in the `config.ru` to remove the conditional logic. That way in the weird case that `falcon` is started in a `test` environment, it still starts. I assume it is here to prevent it running in `rspec`?

benarmston replied:
Well to give us control over whether we want to run it in the specs.

WilliamMcCumstie commented on app/models/job.rb:
```diff
@@ -218,4 +181,41 @@ def validate_batch_script
+    else
+      # They must all be completed then.
+      self.state = 'COMPLETED'
+    end
```
This conditional inherently assumes the only three terminal states are `FAILED`, `CANCELLED`, and `COMPLETED`. This probably fine for the time being, however we will need to remember to update it if and when a new terminal state is added.

An alternative is to preform something along these lines:
```
states  = task.map(&:state).uniq
self.state = TERMINAL_STATES.first { |s| states.include?(s) }
```

This assumes that `TERMINAL_STATES` in precedence order, but that isn't to hard to maintain.

benarmston replied:
That sounds very implicit, and the kind of thing that is easy to get wrong if/when we refactor to use a state machine to handle the job's state.  I'll add a comment above `TERMINAL_STATES` about keeping the two in sync.

WilliamMcCumstie commented on lib/flight_scheduler/array_task_generator.rb:
```diff
@@ -0,0 +1,95 @@
+
+  private
+
+  def get_next_task
```
:panda_face: Generally `get_*` methods should be avoided. As it is private, it doesn't particularly matter.

An alternative would be something like:

```
  def advance_next_task
    @next_task = nil
  end

def next_task
  @next_task ||= task_enum.next
rescue StopIteration
  @next_task = nil
end
```

benarmston replied:
There's going to be refactoring of this class when it needs to support serialization/deserialization.  I'll consider this at that point.

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/fifo_scheduler.rb:
```diff
@@ -28,90 +28,12 @@
-      end.flatten
-    end
-  end
+  class InvalidJobType < RuntimeError; end
```
:panda_face: We should probably look at condensing all the errors into a single inheritance hierarchy at some point, but not a huge priority right now

benarmston replied:
There's a lot to be said for namespacing exceptions inside the classes that throw them.  For instance `JobRegistry::DuplicateJob` reads nicely and better than any non-namespaced name I can think of does.

Having them all inherit from a common base class could be done without changing any namespacing.

But not one for now.

WilliamMcCumstie reviewed (approved):
Few more pedantry things, but otherwise LGTM. They can all probably be addressed at a latter date

benarmston committed:
Comment TERMINAL_STATES/update_array_job_state dependency

benarmston commented:
Thanks for looking at this @WilliamMcCumstie.  I'm quite pleased with how `FifoScheduler` is looking now; moving more code into `allocate_job` and the `candiate` detection block were definitely the right things to do.
